docker/docker_client.go

```diff
@@ -471,7 +473,55 @@ func (c *dockerClient) makeRequest(ctx context.Context, method, path string, hea
 	if err != nil {
 		return nil, err
 	}
-	return c.makeRequestToResolvedURL(ctx, method, url, headers, stream, -1, auth, extraScope)
+	resp, err := c.makeRequestToResolvedURL(ctx, method, url, headers, stream, -1, auth, extraScope)
+
+	// As by default the challenges from the /v2/ "ping" response were used
+	// for fetching the auth token, we might be getting an error back
+	// indicating that the auth scope was insufficient
+	// check for that and update the client challenges to retry with
+	// requesting a new token
+	if retry, newScope := needsRetryWithUpdatedScope(err, resp); retry {
```

Comment:
Is `stream` actually non-nil in any case? I haven't found a single usage of `makeRequest` where stream would not be `nil`

Reply:
At least https://github.com/containers/image/blob/2a56e496eb8637d1c73c26b6da5801a90d09bdb2/docker/docker_image_dest.go#L455 and one other place.

But that’s an interesting point — the most important streaming caller in `PutBlobWithOptions` is not actually calling `makeRequest`:
- Right now, `makeRequest` callers only provide a `bytes.Reader`, so if we replaced the `stream` with a `[]byte`, we could restart the request
- More importantly, conceptually this logic of updating the scope is really a better fit for `makeRequestToResolvedURL`, which is the “more general” request function, rather than `makeRequest`, which is more of a convenience wrapper. (I realize that doesn’t make a difference for the search case.)

Reply:
I have moved the retry logic into `makeRequestToResolvedURL`.

About the `stream` becoming a `[]byte`, can we move this to another PR/issue? I am not overly confident with the `c/image` code base yet and don't want to put too many things into one PR.

Reply:
Absolutely, let‘s _not_ do the `[]byte` change in this PR.

We don’t really have a reason to think that it is going to be true long-term that the `makeRequest` callers will only ever be calling it with a restartable `[]byte` input, so I don’t think we should be restructuring the code for that if it doesn’t buy us anything — and after the move of this scope update inside it doesn’t buy us anything any more.